docs/guide.md

```diff
@@ -0,0 +1,69 @@
+
+# Guide to the Dafny Core Library
+
+_THIS DOCUMENT IS A DRAFT UNDER ACTIVE DEBATE AND EDITING_
+
+A programming language library encapsulates a useful set of basic
+functionality that programmers can then incorporate in their custom
+program without having to rewrite, debug, or verify.
+
+Like other languages, the Dafny programming language and tool set includes
+libraries that provide functionality in various domains. This document describes 
+the _Dafny Core Library_. Dafny programs are copmponed of modules; to ensure
```

Comment:
As I've mentioned offline, I'd like to aim for more, smaller libraries in general, so I'm nervous about committing to a single core library with all the code that's currently here or gets added in the future. But I think everything you've written applies to any library, so perhaps we can keep basically all of it but just say that we are currently considering it one big library, but may break it up in the future?

Reply:
Typo: _copmponed_ ?

Reply:
Fixed the typo. Robin - I can't yet envision the workflow you envision for "lots of little libraries". Can you say a bit more about that?

Reply:
I'm not in a good position to lay out the whole vision yet as I haven't had the time to refine it. :) I'll try to phrase my worries more concretely though.

To date, the `dafny-lang/libraries` (not "library"!) repo has only been consumed by using a submodule and `include`-ing whichever files you need. There hasn't been any concrete attempt to draw sensible boxes around individual "libraries". One of my worries is that if we go the easy route and call the whole thing one library, that will inevitably lead to the runtime footprint of a typical Dafny program getting larger and larger over time, and some target runtimes like Javascript care a lot about this footprint.

I also worry about making "the Core library" too special. If it's the only library we regression test as part of Dafny development, for example, it creates an evolutionary advantage for any code that lives here, and therefore any Dafny developer out there who wants to provide shared code will be incentivized to try to get it into the Core library, which makes Dafny maintainers a bottleneck. I'd be much happier to encourage a community and ecosystem for shared code to be created by anyone and good quality libraries to gain adoption naturally. It's also another case of "drinking our own champagne" if we develop libraries just like end users will.

All of this is ultimately just a nit about the "Dafny Core Library" label, because I don't think I want such a catch-all core library to exist long-term. But perhaps I'd be happy with just a better label for this repo as a collection of libraries.

Reply:
Part of the meaning of "library" (singular or plural) is that you use what you need and not all the rest. Standard linking practice in producing an executable does just that -- only pulls in the material that the user (recursively) needs.
So we need to design the same structure for Dafny -- a means to lazily read and compile the source text as it is needed, not all at once at the beginning -- making sure that modules are no bigger than they need to be. That is a requirement whether the code is packaged as one big thing or many smaller things.

I'm happy to regression test anything we think is stable and wanted enough to expend the resources to watch it and fix it. I don't see this code as special other than that it is what we have.

As for others contributing -- well its open source and they are welcome to. We can state who the owner/maintainer is and when GitHub actions complain they can fix it.

I'm happy to adopt other intuitively-clear names. Just keep in mind that given Dafny's module dependency restrictions, some things are going to be leaves and others not.

Reply:
What you're saying makes sense, but what worries me is assuming that every context that uses a single large library has the means to trim away what isn't used. Given Dafny compiles to many different source languages, and supports separate compilation, that doesn't seem like a safe assumption. Perhaps it's possible for Dafny itself to provide this feature eventually, but that seems like a more complicated solution than just delineating multiple smaller libraries in the first place.

To make it concrete, I was intending for us to just publish the content of `Wrappers.dfy` as a pre-compiled NuGet package, for applications like the Dafny ESDK to use. If we're calling this whole collection of code "the core library", that feel wrong and confusing.

Reply:
You are right that if there is no way to selectively "link" then the user has to do it manually. As a user I would find that a poor UX and annoying. But Dafny already constructs a module dependency graph, so omitting non-dependent modules from compilation does not seem like too much of a stretch. I could imagine even finer dependency tracking, but that would be more work.

Fine to publish Wrappers/Boxes as a start. Though to me it feels too small for the overhead.

Reply:
> omitting non-dependent modules from compilation does not seem like too much of a stretch.

Ah yes, but that can only be done if you're compiling for a specific consumer, not if you're pre-compiling a library for arbitrary consumers.

If you're fine with the fact that we may publish subsets of "the core library" separately in the near future, whatever the meaning of "publish" is, I won't object to the naming for now.